### PR TITLE
OSGI code coverage improvement

### PR DESCRIPTION
**Pull request by serkan-ozal**

### CODE REVIEW

**Comment by mesutcelik** [comment]
It seems related tests have failed.

```
com.hazelcast.osgi.impl.HazelcastOSGiScriptEngineFactoryTest.registerAndGetScriptEngineByExtensionSuccessfully
com.hazelcast.osgi.impl.HazelcastOSGiScriptEngineFactoryTest.registerAndGetScriptEngineByNameSuccessfully
com.hazelcast.osgi.impl.HazelcastOSGiScriptEngineFactoryTest.registerAndGetScriptEngineByMimeTypeSuccessfully
com.hazelcast.osgi.impl.HazelcastOSGiScriptEngineTest.scriptEngineEvaluatedSuccessfully
```

**Comment by serkan-ozal** [comment]
@mesutcelik Yes, I know the reason. I am on it

File: hazelcast/src/main/java/com/hazelcast/osgi/impl/Activator.java [hunk 1]
```diff
@@ -65,18 +65,17 @@ public void stop(BundleContext context)
...
+        } else {
             LOGGER.warning("javax.scripting is not available, scripts from Management Center cannot be executed!");
-            return;
         }
-
-        Class<?> clazz = context.getBundle().loadClass("com.hazelcast.osgi.impl.ScriptEngineActivator");
-        Method register = clazz.getDeclaredMethod("registerOsgiScriptEngineManager", BundleContext.class);
-        register.setAccessible(true);
-        register.invoke(clazz, context);
     }
```
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/osgi/impl/Activator.java]: what is the motivation of restructuring if/else block? to make it more readable?

**Comment by cangencer** [comment]
:+1 looks good

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java [hunk 2]
```diff
@@ -132,6 +206,64 @@ public void defaultInstanceExistWhenItIsSpecified() throws BundleException {
...
+            testBundle = new TestBundle(registerDeregisterListener);
+            try {
+                testBundle.start();
+                fail("OSGI service could not be activated because of exception while registering default instance. " +
+                     "It is expected to get `IllegalStateException` here!");
+            } catch (IllegalStateException e) {
+                // Since bundle is not active, it is expected to get `IllegalStateException`
+            }
+        } finally {
+            if (propValue != null) {
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: Is it not better to put this in a @After method? So that it always gets executed + your tests doesn't become so long.

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java [hunk 3]
```diff
@@ -347,6 +510,30 @@ public void instanceShutdownSuccessfully() {
...
+
+        HazelcastInternalOSGiService service = getService();
+
+        Config config = new Config(INSTANCE_NAME);
+
+        HazelcastOSGiInstance osgiInstance = service.newHazelcastInstance(config);
+        assertTrue(osgiInstance.getLifecycleService().isRunning());
+
+        HazelcastInstance instance = osgiInstance.getDelegatedInstance();
+        assertTrue(instance.getLifecycleService().isRunning());
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: This should not be here. If you don't believe that your instance is running, you should make a test for that explicitly. 

This test should only keep the last 4 asserts.

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java [hunk 4]
```diff
@@ -374,4 +561,62 @@ public void allInstancesShutdownSuccessfully() {
...
+
+        registerDeregisterListener.throwExceptionOnDeregister = true;
+
+        HazelcastInternalOSGiService service = getService();
+
+        osgiInstances.add(service.newHazelcastInstance());
+        osgiInstances.add(service.newHazelcastInstance(new Config("test-osgi-instance")));
+
+        Set<HazelcastOSGiInstance> allOSGiInstances = service.getAllHazelcastInstances();
+        for (HazelcastOSGiInstance osgiInstance : allOSGiInstances) {
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: This should not be here for above reasons.

**Comment by pveentjer** [comment]
Looks good. Minor test comments; but I'll give :+1:

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java [hunk 5]
```diff
@@ -24,17 +28,86 @@
...
+        public void onDeregister(String clazz, TestServiceReference serviceReference) {
+            if (throwExceptionOnDeregister) {
+                if (className != null) {
+                    if (clazz.equals(className)) {
+                        throw new IllegalStateException("You cannot deregister!");
+                    }
+                } else {
+                    throw new IllegalStateException("You cannot deregister!");
+                }
+            }
```
Comment (mesutcelik) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]:
```
 if (!throwExceptionOnDeregister) {
           throw new IllegalStateException("You cannot deregister!");
     }
    if (className != null && clazz.equals(className) {
           throw new IllegalStateException("You cannot deregister!");
     }
```

I would write it this way....but that is just styling...

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java [hunk 6]
```diff
@@ -24,17 +28,86 @@
...
+
+    @BeforeClass
+    public static void beforeClass() {
+        loggingType = System.getProperty("hazelcast.logging.type");
+        Logger rootLogger = Logger.getRootLogger();
+        if ("log4j".equals(loggingType)) {
+            log4jLogLevel = rootLogger.getLevel();
+        }
+        rootLogger.setLevel(Level.TRACE);
+        System.setProperty("hazelcast.logging.type", "log4j");
```
Comment (mesutcelik) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: Do we need this logging hacks in `@AfterClass` and `@BeforeClass`? Did you use it only while developing test cases? if so, we are safe to remove them...

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java [hunk 7]
```diff
@@ -132,6 +206,64 @@ public void defaultInstanceExistWhenItIsSpecified() throws BundleException {
...
+            TestBundleRegisterDeregisterListener registerDeregisterListener =
+                    new TestBundleRegisterDeregisterListener(HazelcastInstance.class.getName());
+            registerDeregisterListener.throwExceptionOnRegister = true;
+
+            testBundle = new TestBundle(registerDeregisterListener);
+            try {
+                testBundle.start();
+                fail("OSGI service could not be activated because of exception while registering default instance. " +
+                     "It is expected to get `IllegalStateException` here!");
+            } catch (IllegalStateException e) {
```
Comment (mesutcelik) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: It can be more readable to use `expected` in the `@Test` annotation instead of `fail`+ `empty catch block`

Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: I made it to be sure that `IllegalStateException` is thrown only there not any other place

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java (hunk 2 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: Restoring properties is done only by some tests not by all of them and the restoring properties are different from test to test. So I though that this approach is less complex.

File: hazelcast/src/main/java/com/hazelcast/osgi/impl/Activator.java (hunk 1 above)
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/osgi/impl/Activator.java]: code coverage :)

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java (hunk 5 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: hmm, maybe. While I am writing, using two nested `if` seemed more readable. I am ok with both of these approaches.

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java (hunk 6 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: for improving code coverage and yes they can be removed safely

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java (hunk 3 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: ok

File: hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java (hunk 4 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/osgi/HazelcastOSGiServiceTest.java]: ok

**Comment by serkan-ozal** [comment]
@pveentjer @mesutcelik are you ok merging this PR now and I will update the tests with direct commit?

**Comment by mesutcelik** [comment]
np for me!
